src/mito2/src/compaction.rs

```diff
@@ -453,31 +454,39 @@ pub struct SerializedCompactionOutput {
     output_time_range: Option<TimestampRange>,
 }
 
-/// Builds [BoxedBatchReader] that reads all SST files and yields batches in primary key order.
-async fn build_sst_reader(
+/// Builders to create [BoxedBatchReader] for compaction.
+struct CompactionSstReaderBuilder<'a> {
     metadata: RegionMetadataRef,
     sst_layer: AccessLayerRef,
     cache: Option<CacheManagerRef>,
-    inputs: &[FileHandle],
+    inputs: &'a [FileHandle],
     append_mode: bool,
     filter_deleted: bool,
     time_range: Option<TimestampRange>,
-) -> Result<BoxedBatchReader> {
-    let mut scan_input = ScanInput::new(sst_layer, ProjectionMapper::all(&metadata)?)
-        .with_files(inputs.to_vec())
-        .with_append_mode(append_mode)
-        .with_cache(cache)
-        .with_filter_deleted(filter_deleted)
-        // We ignore file not found error during compaction.
-        .with_ignore_file_not_found(true);
-
-    // This serves as a workaround of https://github.com/GreptimeTeam/greptimedb/issues/3944
-    // by converting time ranges into predicate.
-    if let Some(time_range) = time_range {
-        scan_input = scan_input.with_predicate(time_range_to_predicate(time_range, &metadata)?);
-    }
+    merge_mode: MergeMode,
+}
 
-    SeqScan::new(scan_input).build_reader().await
+impl<'a> CompactionSstReaderBuilder<'a> {
+    /// Builds [BoxedBatchReader] that reads all SST files and yields batches in primary key order.
+    async fn build_sst_reader(self) -> Result<BoxedBatchReader> {
+        let mut scan_input = ScanInput::new(self.sst_layer, ProjectionMapper::all(&self.metadata)?)
+            .with_files(self.inputs.to_vec())
+            .with_append_mode(self.append_mode)
+            .with_cache(self.cache)
+            .with_filter_deleted(self.filter_deleted)
+            // We ignore file not found error during compaction.
+            .with_ignore_file_not_found(true)
+            .with_merge_mode(self.merge_mode);
+
+        // This serves as a workaround of https://github.com/GreptimeTeam/greptimedb/issues/3944
```

Comment:
BTW. How can we fix this workaround? @v0y4g3r

Reply:
> BTW. How can we fix this workaround? @v0y4g3r

Adding a page filter in current parquet reader to prune pages should be enough. But we need to evaluate the performance in that timestamps in pages is not always monotonic, page filter is fast, but it may not always work. I'll reopen that issue once I have a conclusion.